be/src/fs/fs_s3.cpp

```diff
@@ -111,13 +115,55 @@ class S3ClientFactory {
 
 S3ClientFactory::S3ClientFactory() : _rand((int)::time(nullptr)) {}
 
+// Get a base AWSCredentialsProvider, we can assume role based on this credentials provider
+static std::shared_ptr<Aws::Auth::AWSCredentialsProvider> get_base_aws_credentials_provider(
+        const TAWSCloudCredential& aws_cloud_credential) {
+    if (aws_cloud_credential.use_instance_profile) {
+        return std::make_shared<Aws::Auth::InstanceProfileCredentialsProvider>();
+    } else if (aws_cloud_credential.__isset.access_key && aws_cloud_credential.__isset.secret_key) {
+        return std::make_shared<Aws::Auth::SimpleAWSCredentialsProvider>(aws_cloud_credential.access_key,
+                                                                         aws_cloud_credential.secret_key);
+    } else {
+        DCHECK(false) << "Unreachable!";
+        return std::make_shared<Aws::Auth::AnonymousAWSCredentialsProvider>();
+    }
+}
+
 S3ClientFactory::S3ClientPtr S3ClientFactory::new_client(const ClientConfiguration& config, const FSOptions& opts) {
     std::lock_guard l(_lock);
 
     for (size_t i = 0; i < _items; i++) {
         if (_configs[i] == config) return _clients[i];
     }
 
+    // If cloud credential is set, we use below branch instead of original logic.
```

Comment:
It's better to extract the credential creation logic as a function. You shouldn't bypass the original logical but unify them.